File: examples/node/simple_transfer.js
#!/usr/bin/env node

/**
 * Simple RGTP file transfer example in Node.js
 * 
 * This example demonstrates basic file transfer using RGTP Node.js bindings.
 * It shows both exposing (serving) and pulling (downloading) files.
 */

const rgtp = require('../../bindings/node');
const fs = require('fs');
const path = require('path');
const { performance } = require('perf_hooks');

// ANSI color codes for better output
const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m'
};

function colorize(color, text) {
  return `${colors[color]}${text}${colors.reset}`;
}

function printUsage() {
  console.log(colorize('cyan', 'RGTP Node.js Example'));
  console.log(colorize('cyan', '===================='));
  console.log(`${colorize('bright', 'RGTP Version:')} ${rgtp.VERSION}\n`);
  
  console.log(colorize('yellow', 'Usage:'));
  console.log(`  ${process.argv[1]} expose <filename>`);
  console.log(`  ${process.argv[1]} pull <host> <port> [output_file]\n`);
  
  console.log(colorize('yellow', 'Examples:'));
  console.log(`  ${process.argv[1]} expose document.pdf`);
  console.log(`  ${process.argv[1]} pull localhost 9999`);
  console.log(`  ${process.argv[1]} pull 192.168.1.100 9999 downloaded.pdf`);
}

async function exposeFile(filename) {
  console.log(colorize('blue', `📤 Exposing file: ${filename}`));
  
  // Check if file exists
  if (!fs.existsSync(filename)) {
    console.error(colorize('red', `❌ File not found: ${filename}`));
    process.exit(1);
  }
  
  const stats = fs.statSync(filename);
  console.log(`   📋 Size: ${rgtp.formatBytes(stats.size)}`);
  
  // Create configuration optimized for LAN
  const config = rgtp.createLANConfig();
  config.port = 9999;
  
  // Add progress tracking
  let lastUpdate = 0;
  config.onProgress = (bytesTransferred, totalBytes) => {
    const now = Date.now();
    if (now - lastUpdate > 500) { // Update every 500ms
      const percent = (bytesTransferred / totalBytes * 100).toFixed(1);
      process.stdout.write(`\r   Progress: ${percent}% (${rgtp.formatBytes(bytesTransferred)}/${rgtp.formatBytes(totalBytes)})`);
      lastUpdate = now;
    }
  };
  
  config.onError = (errorCode, errorMessage) => {
    console.error(colorize('red', `\n❌ Error ${errorCode}: ${errorMessage}`));
  };
  
  console.log(`\n${colorize('yellow', 'Configuration:')}`);
  console.log(`   • Port: ${config.port}`);
  console.log(`   • Chunk size: ${rgtp.formatBytes(config.chunkSize)}`);
  console.log(`   • Adaptive mode: ${config.adaptiveMode}`);
  console.log(`   • Exposure rate: ${config.exposureRate} chunks/sec`);
  
  // Create session
  const session = new rgtp.Session(config);
  
  // Handle graceful shutdown
  let shutdownRequested = false;
  process.on('SIGINT', async () => {
    if (shutdownRequested) {
      console.log(colorize('red', '\n🛑 Force shutdown'));
      process.exit(1);
    }
    
    shutdownRequested = true;
    console.log(colorize('yellow', '\n🛑 Shutting down gracefully... (Press Ctrl+C again to force)'));
    
    try {
      await session.cancel();
      session.close();
      process.exit(0);
    } catch (error) {
      console.error(colorize('red', `Error during shutdown: ${error.message}`));
      process.exit(1);
    }
  });
  
  try {
    const startTime = performance.now();
    
    // Start exposure
    await session.exposeFile(filename);
    
    console.log(colorize('green', `\n✅ File exposed successfully on port ${config.port}`));
    console.log(`${colorize('cyan', 'Clients can now pull from:')} <this_host>:${config.port}`);
    console.log(colorize('yellow', 'Press Ctrl+C to stop...\n'));
    
    // Monitor progress
    const monitorInterval = setInterval(async () => {
      try {
        const stats = await session.getStats();
        if (stats.chunksTransferred > 0) {
          process.stdout.write(`\r${colorize('cyan', 'Stats:')} ${rgtp.formatBytes(stats.bytesTransferred)} transferred, ${stats.throughputMbps.toFixed(1)} MB/s, ${stats.completionPercent.toFixed(1)}% complete`);
        }
      } catch (error) {
        // Ignore errors during monitoring
      }
    }, 1000);
    
    // Wait for completion
    await session.waitComplete();
    
    clearInterval(monitorInterval);
    const endTime = performance.now();
    
    console.log(colorize('green', '\n✅ Exposure completed successfully!'));
    
    // Print final statistics
    const finalStats = await session.getStats();
    printFinalStats(finalStats, 'Exposure', endTime - startTime);
    
  } catch (error) {
    console.error(colorize('red', `\n❌ Exposure failed: ${error.message}`));
    process.exit(1);
  } finally {
    session.close();
  }
}

async function pullFile(host, port, outputFile) {
  console.log(colorize('blue', `📥 Pulling from ${host}:${port} -> ${outputFile}`));
  
  // Create configuration optimized for WAN
  const config = rgtp.createWANConfig();
  
  // Add progress tracking with progress bar
  let lastUpdate = 0;
  config.onProgress = (bytesTransferred, totalBytes) => {
    const now = Date.now();
    if (now - lastUpdate > 200) { // Update every 200ms
      const percent = bytesTransferred / totalBytes * 100;
      
      // Simple progress bar
      const barWidth = 40;
      const filled = Math.floor(percent / 100 * barWidth);
      
      let bar = '[';
      for (let i = 0; i < barWidth; i++) {
        if (i < filled) {
          bar += '=';
        } else if (i === filled) {
          bar += '>';
        } else {
          bar += ' ';
        }
      }
      bar += ']';
      
      process.stdout.write(`\r   ${bar} ${percent.toFixed(1)}% (${rgtp.formatBytes(bytesTransferred)}/${rgtp.formatBytes(totalBytes)})`);
      lastUpdate = now;
    }
  };
  
  config.onError = (errorCode, errorMessage) => {
    console.error(colorize('red', `\n❌ Error ${errorCode}: ${errorMessage}`));
  };
  
  console.log(`\n${colorize('yellow', 'Configuration:')}`);
  console.log(`   • Timeout: ${rgtp.formatDuration(config.timeout)}`);
  console.log(`   • Chunk size: ${rgtp.formatBytes(config.chunkSize)}`);
  console.log(`   • Adaptive mode: ${config.adaptiveMode}`);
  
  // Create client
  const client = new rgtp.Client(config);
  
  // Handle graceful shutdown
  process.on('SIGINT', async () => {
    console.log(colorize('yellow', '\n🛑 Cancelling pull operation...'));
    try {
      await client.cancel();
      client.close();
      process.exit(0);
    } catch (error) {
      console.error(colorize('red', `Error during cancellation: ${error.message}`));
      process.exit(1);
    }
  });
  
  try {
    console.log('\nStarting pull operation...');
    const startTime = performance.now();
    
    // Pull the file
    await client.pullToFile(host, port, outputFile);
    
    const endTime = performance.now();
    
    console.log(colorize('green', '\n✅ Pull completed successfully!'));
    console.log(`   • File saved as: ${outputFile}`);
    console.log(`   • Total time: ${rgtp.formatDuration(endTime - startTime)}`);
    
    // Print final statistics
    const finalStats = await client.getStats();
    printFinalStats(finalStats, 'Pull', endTime - startTime);
    
  } catch (error) {
    console.error(colorize('red', `\n❌ Pull failed: ${error.message}`));
    
    // Try to get partial statistics
    try {
      const stats = await client.getStats();
      if (stats.bytesTransferred > 0) {
        console.log(`   • Partial data received: ${rgtp.formatBytes(stats.bytesTransferred)} (${stats.completionPercent.toFixed(1)}%)`);
      }
    } catch (statsError) {
      // Ignore stats error
    }
    
    process.exit(1);
  } finally {
    client.close();
  }
}

function printFinalStats(stats, operation, totalTimeMs) {
  console.log(`\n${colorize('cyan', `📊 ${operation} Statistics:`)}`);
  console.log(`   • File size: ${rgtp.formatBytes(stats.totalBytes)}`);
  console.log(`   • Duration: ${rgtp.formatDuration(stats.elapsedMs)}`);
  console.log(`   • Average throughput: ${stats.avgThroughputMbps.toFixed(2)} MB/s`);
  console.log(`   • Peak throughput: ${stats.throughputMbps.toFixed(2)} MB/s`);
  console.log(`   • Chunks transferred: ${stats.chunksTransferred}/${stats.totalChunks}`);
  console.log(`   • Retransmissions: ${stats.retransmissions}`);
  console.log(`   • Efficiency: ${stats.efficiencyPercent.toFixed(1)}%`);
  
  if (stats.totalBytes > 0 && totalTimeMs > 0) {
    const effectiveSpeed = (stats.totalBytes / (1024 * 1024)) / (totalTimeMs / 1000);
    console.log(`   • Effective speed: ${effectiveSpeed.toFixed(2)} MB/s`);
  }
}

async function main() {
  if (process.argv.length < 3) {
    printUsage();
    process.exit(1);
  }
  
  const command = process.argv[2].toLowerCase();
  
  try {
    switch (command) {
      case 'expose':
        if (process.argv.length !== 4) {
          console.error(colorize('red', 'Usage: expose <filename>'));
          process.exit(1);
        }
        await exposeFile(process.argv[3]);
        break;
        
      case 'pull':
        if (process.argv.length < 5 || process.argv.length > 6) {
          console.error(colorize('red', 'Usage: pull <host> <port> [output_file]'));
          process.exit(1);
        }
        
        const host = process.argv[3];
        const port = parseInt(process.argv[4]);
        const outputFile = process.argv[5] || 'rgtp_download.bin';
        
        if (isNaN(port) || port < 1 || port > 65535) {
          console.error(colorize('red', 'Invalid port number'));
          process.exit(1);
        }
        
        await pullFile(host, port, outputFile);
        break;
        
      default:
        console.error(colorize('red', `Unknown command: ${command}`));
        printUsage();
        process.exit(1);
    }
  } catch (error) {
    console.error(colorize('red', `💥 Unexpected error: ${error.message}`));
    if (process.env.NODE_ENV === 'development') {
      console.error(error.stack);
    }
    process.exit(1);
  }
}

// Run the main function
if (require.main === module) {
  main().catch(error => {
    console.error(colorize('red', `Fatal error: ${error.message}`));
    process.exit(1);
  });
}

File: examples/go/simple_transfer.go
// Simple RGTP file transfer example in Go
//
// This example demonstrates basic file transfer using RGTP Go bindings.
// It shows both exposing (serving) and pulling (downloading) files.

package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"os/signal"
	"path/filepath"
	"strconv"
	"syscall"
	"time"

	"github.com/red-giant-protocol/rgtp-go"
)

func main() {
	fmt.Println("RGTP Go Example")
	fmt.Println("===============")
	fmt.Printf("RGTP Version: %s\n\n", rgtp.Version())

	if len(os.Args) < 2 {
		printUsage()
		os.Exit(1)
	}

	command := os.Args[1]

	switch command {
	case "expose":
		if len(os.Args) != 3 {
			fmt.Println("Usage: expose <filename>")
			os.Exit(1)
		}
		filename := os.Args[2]
		exposeFile(filename)

	case "pull":
		if len(os.Args) < 4 || len(os.Args) > 5 {
			fmt.Println("Usage: pull <host> <port> [output_file]")
			os.Exit(1)
		}
		host := os.Args[2]
		port, err := strconv.ParseUint(os.Args[3], 10, 16)
		if err != nil {
			log.Fatalf("Invalid port: %v", err)
		}
		outputFile := "rgtp_download.bin"
		if len(os.Args) == 5 {
			outputFile = os.Args[4]
		}
		pullFile(host, uint16(port), outputFile)

	default:
		fmt.Printf("Unknown command: %s\n", command)
		printUsage()
		os.Exit(1)
	}
}

func printUsage() {
	fmt.Println("Usage:")
	fmt.Printf("  %s expose <filename>\n", os.Args[0])
	fmt.Printf("  %s pull <host> <port> [output_file]\n", os.Args[0])
	fmt.Println()
	fmt.Println("Examples:")
	fmt.Printf("  %s expose document.pdf\n", os.Args[0])
	fmt.Printf("  %s pull localhost 9999\n", os.Args[0])
	fmt.Printf("  %s pull 192.168.1.100 9999 downloaded.pdf\n", os.Args[0])
}

func exposeFile(filename string) {
	fmt.Printf("📤 Exposing file: %s\n", filename)

	// Check if file exists
	if _, err := os.Stat(filename); os.IsNotExist(err) {
		log.Fatalf("File not found: %s", filename)
	}

	// Create configuration optimized for LAN
	config := rgtp.LANConfig()
	config.Port = 9999
	config.ProgressCallback = func(bytesTransferred, totalBytes uint64) {
		if totalBytes > 0 {
			percent := float64(bytesTransferred) / float64(totalBytes) * 100.0
			fmt.Printf("\rProgress: %.1f%% (%s/%s)", 
				percent, 
				formatBytes(bytesTransferred), 
				formatBytes(totalBytes))
		}
	}
	config.ErrorCallback = func(errorCode int, errorMessage string) {
		fmt.Printf("\nError %d: %s\n", errorCode, errorMessage)
	}

	fmt.Printf("Configuration:\n")
	fmt.Printf("  • Port: %d\n", config.Port)
	fmt.Printf("  • Chunk size: %s\n", formatBytes(uint64(config.ChunkSize)))
	fmt.Printf("  • Adaptive mode: %v\n", config.AdaptiveMode)
	fmt.Printf("  • Exposure rate: %d chunks/sec\n", config.ExposureRate)

	// Create session
	session, err := rgtp.NewSession(config)
	if err != nil {
		log.Fatalf("Failed to create session: %v", err)
	}
	defer session.Close()

	// Set up signal handling for graceful shutdown
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	// Start exposure in a goroutine
	done := make(chan error, 1)
	go func() {
		err := session.ExposeFile(filename)
		if err != nil {
			done <- err
			return
		}

		fmt.Printf("\n✅ File exposed successfully on port %d\n", config.Port)
		fmt.Println("Clients can now pull from: <this_host>:9999")
		fmt.Println("Press Ctrl+C to stop...\n")

		err = session.WaitComplete()
		done <- err
	}()

	// Monitor progress and handle shutdown
	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			stats, err := session.GetStats()
			if err == nil && stats.ChunksTransferred > 0 {
				fmt.Printf("\rStats: %s transferred, %.1f MB/s, %.1f%% complete",
					formatBytes(stats.BytesTransferred),
					stats.ThroughputMbps,
					stats.CompletionPercent)
			}

		case <-sigChan:
			fmt.Println("\n🛑 Shutting down...")
			session.Cancel()
			return

		case err := <-done:
			if err != nil {
				fmt.Printf("\n❌ Exposure failed: %v\n", err)
				os.Exit(1)
			}
			fmt.Println("\n✅ Exposure completed successfully!")
			
			// Print final statistics
			if stats, err := session.GetStats(); err == nil {
				printFinalStats(stats, "Exposure")
			}
			return
		}
	}
}

func pullFile(host string, port uint16, outputFile string) {
	fmt.Printf("📥 Pulling from %s:%d -> %s\n", host, port, outputFile)

	// Create configuration optimized for WAN
	config := rgtp.WANConfig()
	config.ProgressCallback = func(bytesTransferred, totalBytes uint64) {
		if totalBytes > 0 {
			percent := float64(bytesTransferred) / float64(totalBytes) * 100.0
			
			// Simple progress bar
			barWidth := 50
			filled := int(percent / 100.0 * float64(barWidth))
			
			fmt.Printf("\r[")
			for i := 0; i < barWidth; i++ {
				if i < filled {
					fmt.Print("=")
				} else if i == filled {
					fmt.Print(">")
				} else {
					fmt.Print(" ")
				}
			}
			fmt.Printf("] %.1f%% (%s/%s)", 
				percent, 
				formatBytes(bytesTransferred), 
				formatBytes(totalBytes))
		}
	}
	config.ErrorCallback = func(errorCode int, errorMessage string) {
		fmt.Printf("\nError %d: %s\n", errorCode, errorMessage)
	}

	fmt.Printf("Configuration:\n")
	fmt.Printf("  • Timeout: %v\n", config.Timeout)
	fmt.Printf("  • Chunk size: %s\n", formatBytes(uint64(config.ChunkSize)))
	fmt.Printf("  • Adaptive mode: %v\n", config.AdaptiveMode)

	// Create client
	client, err := rgtp.NewClient(config)
	if err != nil {
		log.Fatalf("Failed to create client: %v", err)
	}
	defer client.Close()

	fmt.Println("Starting pull operation...")

	// Record start time
	startTime := time.Now()

	// Pull the file
	err = client.PullToFile(host, port, outputFile)
	
	// Calculate elapsed time
	elapsed := time.Since(startTime)

	if err != nil {
		fmt.Printf("\n❌ Pull failed: %v\n", err)
		
		// Try to get partial statistics
		if stats, err := client.GetStats(); err == nil && stats.BytesTransferred > 0 {
			fmt.Printf("   • Partial data received: %s (%.1f%%)\n",
				formatBytes(stats.BytesTransferred),
				stats.CompletionPercent)
		}
		os.Exit(1)
	}

	fmt.Printf("\n✅ Pull completed successfully!\n")
	fmt.Printf("   • File saved as: %s\n", outputFile)
	fmt.Printf("   • Total time: %v\n", elapsed)

	// Get final statistics
	if stats, err := client.GetStats(); err == nil {
		printFinalStats(stats, "Pull")
	}
}

func printFinalStats(stats *rgtp.Stats, operation string) {
	fmt.Printf("\n📊 %s Statistics:\n", operation)
	fmt.Printf("   • File size: %s\n", formatBytes(stats.TotalBytes))
	fmt.Printf("   • Duration: %v\n", stats.Elapsed)
	fmt.Printf("   • Average throughput: %.2f MB/s\n", stats.AvgThroughputMbps)
	fmt.Printf("   • Peak throughput: %.2f MB/s\n", stats.ThroughputMbps)
	fmt.Printf("   • Chunks transferred: %d/%d\n", stats.ChunksTransferred, stats.TotalChunks)
	fmt.Printf("   • Retransmissions: %d\n", stats.Retransmissions)
	fmt.Printf("   • Efficiency: %.1f%%\n", stats.EfficiencyPercent())
	
	if stats.TotalBytes > 0 && stats.Elapsed > 0 {
		effectiveSpeed := float64(stats.TotalBytes) / (1024 * 1024) / stats.Elapsed.Seconds()
		fmt.Printf("   • Effective speed: %.2f MB/s\n", effectiveSpeed)
	}
}

func formatBytes(bytes uint64) string {
	const unit = 1024
	if bytes < unit {
		return fmt.Sprintf("%d B", bytes)
	}
	div, exp := uint64(unit), 0
	for n := bytes / unit; n >= unit; n /= unit {
		div *= unit
		exp++
	}
	return fmt.Sprintf("%.1f %cB", float64(bytes)/float64(div), "KMGTPE"[exp])
}